Soft wrap hanging indentation spaces

as-cii opened the pull request:
Closes https://github.com/atom/atom/issues/6003

/cc: @nathansobo

as-cii committed:
Add soft-wrap hanging indentation spaces

as-cii committed:
:white_check_mark: Verify that hanging indentation is tokenized

as-cii commented on spec/display-buffer-spec.coffee:
```diff
@@ -128,6 +128,12 @@ describe "DisplayBuffer", ->
           expect(displayBuffer.tokenizedLineForScreenRow(4).tokens[0].isSoftWrapIndentation).toBeTruthy()
           expect(displayBuffer.tokenizedLineForScreenRow(4).tokens[1].isSoftWrapIndentation).toBeTruthy()
 
+        it "correctly tokenizes hanging indentation spaces", ->
```
Please note that I wrote only a test that verifies the correct tokenization of `hangingIndentationSpaces`, as positioning was already thoroughly tested by other specs. I believe that adding similar ones for these hanging spaces was redundant and not worth the maintenance.

as-cii committed:
Add default editor.softWrapHangingIndentationSpaces

as-cii committed:
:white_check_mark: Check leading spaces as well

thomasjo commented:
Bloody brilliant! :sparkling_heart:

nathansobo commented on src/tokenized-line.coffee:
```diff
@@ -118,20 +118,27 @@ class TokenizedLine
     Math.round(@tabLength * oddIndentLevel)
 
-  buildSoftWrapIndentationTokens: (token) ->
+  buildSoftWrapIndentationTokens: (token, hangingIndentationSpaces) ->
```
What's your reasoning behind breaking the "hanging" indentation into a separate token rather than just making the initial indentation token longer?

as-cii replied:
@nathansobo: if I understand your question correctly, we need to break the "hanging spaces" into tokens so that we could show indent guides.

After a bit of sleep (I love Hammock Driven Development :sleeping:), however, I realized this wasn't doing what I expected: indeed indent guides weren't shown, as only one token was created for all the hanging indentation spaces.

f0b9bb7ce3e75f716c0d0297bbcacee5fe289f81 solves this by calculating the `hangingIndentLevel` and summing it to `TokenizedLine#indentLevel`.

nathansobo commented:
Excited about the feature, but I'm wondering if anyone has any ideas for the name of the setting. `softWrapHangingIndentationSpaces` feels a bit wordy. Maybe `softWrapIndentationOverhang`? Not loving that either. We could go with it I'd just like to kick around some alternatives.

as-cii committed:
Show indent guides on hanging indentation too

as-cii commented:
> Excited about the feature, but I'm wondering if anyone has any ideas for the name of the setting. softWrapHangingIndentationSpaces feels a bit wordy. Maybe softWrapIndentationOverhang? Not loving that either. We could go with it I'd just like to kick around some alternatives.

Yeah, naming things is hard :disappointed: To be honest I simply borrowed the term from the referenced pull-request but it feels wordy to me as well. Maybe we could involve @atom/non-github-maintainers to help us with this? Your help would be much appreciated :+1: 

As usual, thank you @nathansobo for the interesting discussion and for your always great feedback! :pray:

as-cii committed:
:memo:

grizzilus commented:
I just stumbled across this PR and the naming conversation.

From the typography perspective I like either the current `softWrapHangingIndentationSpaces` or a shorter `softWrapHangingIndentation`. In typography, the general paragraph shape that results from line-wrapping like that which results from having the option in this PR enabled can come about one of two ways: either the paragraph has an `outdentation` or a `hanging indentation`. (I personally have not seen any other names, and a quick flip through all the typography books I have at hand seems to confirm this.)
 
An `outdentation` is when the paragraph extends into the left margin by some amount. The analogue in a text editor would by a negative-valued indentation width, which doesn't really make sense. But this brings up another interesting concern I'll mention at the end of this comment.
 
A `hanging indentation` is is an indent of all lines after some _n_th line of the paragraph by some amount, which, in my opinion, corresponds exactly to the behavior of this PR. So having `HangingIndentation` somewhere in the naming makes it very clear what is going on. In particular, the original issues was created used `TeX` as an example, where this type of output is controlled by the lengths `\hangindent` and `\hangafter`, which may be why the author of that issue used the term `hanging` in the name of the issue.

The `outdentation` discussion bring up an interesting point: this PR results in some odd behavior if `softWrapHangingIndentationSpaces` is negative and its magnitude is larger than the main indent level. I would expect the soft-wrapped lines to be aligned with the left side of the editor view, but they indent themselves by the difference between the value of `softWrapHangingIndentationSpaces` and the number of spaces by which the first line is indented. (It will be clear if you try some negative values.)

50Wliu commented:
`indentSoftWrap`?  That sounds like it could be a bit misleading though so perhaps a comment could be added stating "this is relative to the original indent" or something?

as-cii committed:
:bug: Hanging indent must be a positive value

as-cii committed:
:art: Rename to softWrapHangingIndent

as-cii commented:
@50Wliu @grizzilus: thank you for answers! 

Based on @grizzilus thorough explanation (thank you for that), I renamed the config variable for now to `softWrapHangingIndent` but, most importantly, we discovered an interesting edge case, which I didn't take into account. As I intended the feature, the value of this config setting **must always be positive** and, therefore, I fixed that accordingly (https://github.com/atom/atom/commit/e5fa44171df82962efb185d6b5d36ab2b8329f86).

Let's wait for @nathansobo final take on this :fireworks:

grizzilus commented:
I'm actually not getting any handing indent if I build the last commit here but I can't find any error (or anything missed) in 7c33b9b...

as-cii commented:
@grizzilus: could you post here your `config.cson`, please? Also, make sure you have updated the setting to:

```
softWrapHangingIndent: <number>
```

grizzilus commented:
@as-cii Nevermind, it is fine, sorry about that. I rebuilt but when changing the value in the GUI it wasn't updating. Removing the old setting from the config file allowed me to set directly or from the GUI.

On a side note, I wonder if it is reasonable to allow for negative indents. I know you originally intended for positive values only, but what if we did allow negative indents and just cap it at the maximum of 0 and the sum of the regular indent and the hanging indent? While I can't think of an really useful example (other than easily seeing where there are long lines nested very deep more easily), it seems like a straightforward change to allow more flexibility.

as-cii commented:
> While I can't think of an really useful example (other than easily seeing where there are long lines nested very deep more easily), it seems like a straightforward change to allow more flexibility.

Thank you for pointing this out @grizzilus. I must say I have considered this option while implementing the change, but as you're pointing out I couldn't find a meaningful example and, therefore, I opted to simply make the configuration stricter allowing only positive values. Supposedly, this makes its intent a bit clearer (or does it?) sacrificing a bit of flexibility (e.g. negative values).

On the other hand, if we find a good use-case for negative values I'll be happy to implement that feature as well. :dancers: 

/cc: @AlexLog @jtkiley

grizzilus commented:
The more I think about it, you are right. I can't imagine the magnitude of this being more than twice the indentation amount, and if text starts that close to the margin, there are bigger issues to deal with than the hanging indent. But great work @as-cii on this and the original indented softwrap; it is such a help to people who use TeX-based files regularly.

paulcbetts commented:
How about `wrapHangingIndentation` - Atom doesn't do hard-wrap, and "IndentationSpaces" seems redundant

jtkiley commented:
@as-cii @grizzilus I can't think of a good use of negative values, either. I thought about it briefly when I was writing up the issue, and I thought about it after seeing the discussion this morning. Someone may come up with a cool use case, but a negative indent would actually look like the old behavior (i.e. wrapped lines beginning at the left margin) in some cases. If anyone is clamoring for the old behavior as an option, I haven't seen it.

Echoing @grizzilus, thank you @as-cii for the awesome work on indented soft wrap and the hanging indent. It's a huge help for TeX work.

nathansobo commented on src/config-schema.coffee:
```diff
@@ -149,6 +149,10 @@ module.exports =
       softWrapAtPreferredLineLength:
         type: 'boolean'
         default: false
+      softWrapHangingIndent:
```
:+1: I like the new shorter name.

nathansobo commented on src/tokenized-line.coffee:
```diff
@@ -111,27 +111,32 @@ class TokenizedLine
       indentTokens
 
-  softWrapAt: (column) ->
+  softWrapAt: (column, options = {}) ->
```
What do you think about just passing the overhang in as a direct parameter rather than using an options object. It's not as pretty but it saves an object allocation and since this code is internal, seems like a reasonable trade-off.

nathansobo commented on spec/display-buffer-spec.coffee:
```diff
@@ -128,6 +128,20 @@ describe "DisplayBuffer", ->
+          atom.config.set('editor.softWrapHangingIndent', 3)
+
+        it "further indents wrapped lines", ->
+          expect(displayBuffer.tokenizedLineForScreenRow(11).text).toBe "       sort(left).concat(pivot).concat(sort(right)"
```
Could you also show the text for row 10 here? I think that would tell the :book: a bit more.

nathansobo commented on spec/display-buffer-spec.coffee:
```diff
@@ -128,6 +128,20 @@ describe "DisplayBuffer", ->
+          expect(displayBuffer.tokenizedLineForScreenRow(11).text).toBe "       sort(left).concat(pivot).concat(sort(right)"
+          expect(displayBuffer.tokenizedLineForScreenRow(12).text).toBe "       );"
+
+        it "correctly tokenizes hanging indentation spaces", ->
```
Some suggestions for clarifying this example:

* Talk about the content of each token.
* Assert that the token at index 4 is *not* soft-wrap indentation to clarify the boundary.
* Maybe mention why we're breaking the indentation up in the description: maybe "breaks soft-wrap indentation into a token for each indentation level to support indent guides".

as-cii committed:
:white_check_mark: Write a more comprehensive spec

as-cii committed:
Avoid named parameter to save an allocation

nathansobo commented on src/tokenized-line.coffee:
```diff
@@ -111,27 +111,32 @@ class TokenizedLine
+  # indentLevel - {Number}
+  # Returns a {Number} representing the odd indentation spaces in `indentLevel`.
+  oddIndentationSpacesForIndentLevel: (indentLevel) ->
+    oddIndentLevel = indentLevel - Math.floor(indentLevel)
```
Thought here: couldn't you also do it this way?

```coffee
(indentLevel * @tabLength) % @tabLength
```

nathansobo replied:
Actually, I have a suggestion for removing this method entirely by changing the implementation of `buildSoftWrapIndentationTokens` below...

nathansobo commented on src/tokenized-line.coffee:
```diff
@@ -111,27 +111,32 @@ class TokenizedLine
 
-  buildSoftWrapIndentationTokens: (token) ->
-    indentTokens = [0...Math.floor(@indentLevel)].map =>
+  buildSoftWrapIndentationTokens: (token, hangingIndent) ->
```
If you change the implementation of this method, you don't need `oddIndentationSpacesForIndentLevel` at all. 

```coffee
  buildSoftWrapIndentationTokens: (token, hangingIndent) ->
    totalIndentSpaces = (@indentLevel * @tabLength) + hangingIndent
    indentTokens = []
    while totalIndentSpaces > 0
      tokenLength = Math.min(@tabLength, totalIndentSpaces)
      indentToken = token.buildSoftWrapIndentationToken(tokenLength)
      indentTokens.push(indentToken)
      totalIndentSpaces -= tokenLength
    indentTokens
```

I tested this version locally, but wanted to avoid committing to this branch without your :+1:.

as-cii committed:
Even more descriptive specs :sparkles:

as-cii committed:
:art: Refactor buildSoftWrapIndentationTokens

...as suggested by @nathansobo :sparkles:

as-cii commented:
@nathansobo: thank you for all the great suggestions: hopefully, I have implemented them all :+1: 

Please, have one final look and feel free to commit on this branch if you need to :smile:

nathansobo commented:
:joy: :joy_cat: Members of every species thank you for your contribution.
